wal-vdom/src/virtual_dom/vnode.rs

```diff
@@ -1,24 +1,172 @@
-use gloo::console::__macro::JsValue;
+use gloo::{console::log, utils::document};
 use serde::Serialize;
-use web_sys::Node;
+use web_sys::{Element, Node, Text};
 
 use super::{VElement, VText};
 
-#[derive(Serialize)]
+#[derive(Debug, Serialize)]
 pub enum VNode {
-    Element{
-        velement: VElement,
+    Element {
+        #[serde(skip_serializing)]
+        concrete: Option<Element>,
+        virt: VElement,
     },
-    Text{
-        vtext: VText,
+    Text {
+        #[serde(skip_serializing)]
+        concrete: Option<Text>,
+        virt: VText,
     },
 }
 
 impl VNode {
-    pub fn render(&self) -> Result<Node, JsValue> {
+    pub fn patch(&mut self, last: Option<VNode>, ancestor: &Node) {
         match self {
-            VNode::Element { velement } => velement.render(),
-            VNode::Text { vtext } => vtext.render(),
+            VNode::Element {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_element(concrete, virt, last, ancestor),
+            VNode::Text {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_text(concrete, virt, last, ancestor),
+        };
+    }
+
+    fn patch_element(
+        concrete: &mut Option<Element>,
+        virt: &mut VElement,
+        last: Option<VNode>,
+        ancestor: &Node,
+    ) {
+        log!("Patching element");
+        let mut old_virt: Option<VElement> = None;
+        match last {
+            // First creating the node
+            Some(VNode::Element { concrete: None, .. })
+            | Some(VNode::Text { concrete: None, .. })
+            | None => {
+                log!("\tCreating the node from the bottom");
+                let new_el = document()
+                    .create_element(&virt.tag_name)
+                    .expect("Couldnt create new element");
+                ancestor
+                    .append_child(&new_el)
+                    .expect("Couldnt append child to node");
+                *concrete = Some(new_el);
+            }
+
+            // Just copy
+            Some(VNode::Element { concrete: Some(el), virt, }) => {
+                log!("\tCopying existing node");
+                *concrete = Some(el);
+                old_virt = Some(virt);
+            }
+
+            // Swap nodes
+            Some(VNode::Text { concrete: Some(text), .. }) => {
+                log!("\tSwaping existing node");
+                let new_el = document()
+                    .create_element(&virt.tag_name)
+                    .expect("Couldnt create new element");
+                ancestor
+                    .replace_child(&text, &new_el)
+                    .expect("Couldnt replace child with a new node");
+                *concrete = Some(new_el);
+            }
+        };
+
+        // Render over concrete new element
+        let target = concrete.as_mut().expect("It shouldnt be none");
```

Comment:
wydaje sie ze error message w wielu miejscach mogłyby dawać więcej info, a w takiej sytuacji gdy powyższy kod powinien zawsze przypisać coś do concrete to imo nawet unwrap by nie zaszkodził